lib/count_overdraft_events/main.py

```diff
@@ -1,77 +1,92 @@
 #!/usr/bin/env python3
-
+import asyncio
 import os
 from datetime import datetime, timedelta
 
-from pngme.api import Client
+import pandas as pd  # type: ignore
+from typing import Tuple
 
+from pngme.api import AsyncClient
 
-def get_count_overdraft_events(
-    api_client: Client, user_uuid: str, utc_starttime: datetime, utc_endtime: datetime
-) -> float:
-    """
-    Count events labelled with Overdraft across all institutions.
 
-    Typical date ranges are last 30 days, 31-60 days and 61-90 days.
+async def get_count_overdraft_events(
+    api_client: AsyncClient, user_uuid: str, utc_time: datetime
+) -> Tuple[int, int, int]:
+    """
+    Count events labelled with Overdraft across all institutions
+    over the following date ranges: last 30 days, 31-60 days and 61-90 days.
 
     Args:
-        api_client: Pngme API client
+        api_client: Pngme Async API client
         user_uuid: the Pngme user_uuid for the mobile phone user
-        utc_starttime: the datetime for the left-hand-side of the time-window
-        utc_endtime: the datetime for the right-hand-side of the time-window
+        utc_time: the time-zero to use in constructing the 0-30, 31-60 and 61-90 windows
 
     Returns:
         count of overdraft events within the given time window
     """
     label = "Overdraft"
 
-    institutions = api_client.institutions.get(user_uuid)
+    institutions = await api_client.institutions.get(user_uuid=user_uuid)
 
-    overdraft_count = 0
-    for institution in institutions:
-        overdraft_events = api_client.alerts.get(
+    utc_starttime = utc_time - timedelta(days=90)
+    inst_coroutines = [
+        api_client.alerts.get(
             user_uuid=user_uuid,
             institution_id=institution.institution_id,
             utc_starttime=utc_starttime,
-            utc_endtime=utc_endtime,
+            utc_endtime=utc_time,
             labels=[label],
         )
-        overdraft_count += len(overdraft_events)
+        for institution in institutions
+    ]
+
+    r = await asyncio.gather(*inst_coroutines)
+
+    record_list = []
+    for inst_list in r:
+        record_list.extend([dict(alert) for alert in inst_list])
+
+    # if no data available for the user, assume count of overdraft event is zero
+    if len(record_list) == 0:
+        return 0, 0, 0
+
+    record_df = pd.DataFrame(record_list)
 
-    return overdraft_count
+    # Get the total inbound credit over a period
+    ts_30 = (utc_time - timedelta(days=30)).timestamp()
+    ts_60 = (utc_time - timedelta(days=60)).timestamp()
+    ts_90 = (utc_time - timedelta(days=90)).timestamp()
+
+    coe_0_30 = 0
+    coe_31_60 = 0
+    coe_61_90 = 0
+
+    for _, row in record_df.iterrows():
```

Comment:
can you refactor this to be more data-sciency instead of swe'y?
if we're pulling in pandas, we can use pandas to do the summation.
something like:
```
coe_31_60 = len(record_df[record_df["ts"] <  ts_30 & record_df["ts"] <  ts_30])
```

Reply:
Done ✅